d2game/d2player/quest_log.go

```diff
@@ -235,131 +223,200 @@ func (s *QuestLog) Load() {
 
 	var questsInAct int
-	if act == act4 {
-		questsInAct = act4QuestsNumber
+	if act == d2enum.Act4 {
+		questsInAct = d2enum.HalfQuestsNumber
 	} else {
-		questsInAct = normalActQuestsNumber
+		questsInAct = d2enum.NormalActQuestsNumber
 	}
 
-	for n := 0; n < questsInAct; n++ {
-		q := s.questTable(act, n)
+	var sockets []*d2ui.Sprite
 
-		button := s.uiManager.NewButton(d2ui.ButtonTypeBlankQuestBtn, "")
-		button.SetPosition(q.x+questOffsetX, q.y+questOffsetY)
-		button.OnActivated(s.makeQuestCallback(n))
+	var buttons []*d2ui.Button
+
+	var icon *d2ui.Sprite
+
+	for n := 0; n < questsInAct; n++ {
+		x, y := s.getPositionForSocket(n)
 
 		socket, err := s.uiManager.NewSprite(d2resource.QuestLogSocket, d2resource.PaletteSky)
 		if err != nil {
 			s.Error(err.Error())
 		}
 
-		socket.SetPosition(q.x+questOffsetX, q.y+iconOffsetY+2*questOffsetY)
+		socket.SetPosition(x+questOffsetX, y+iconOffsetY+2*questOffsetY)
+		sockets = append(sockets, socket)
 
-		icon, err := s.uiManager.NewSprite(d2resource.QuestLogDone, d2resource.PaletteSky)
-		if err != nil {
-			s.Error(err.Error())
-		}
+		button := s.uiManager.NewButton(d2ui.ButtonTypeBlankQuestBtn, "")
+		button.SetPosition(x+questOffsetX, y+questOffsetY)
+		buttons = append(buttons, button)
 
-		err = icon.SetCurrentFrame(q.frame)
+		icon, err = s.makeQuestIconForAct(act, n)
 		if err != nil {
 			s.Error(err.Error())
 		}
 
-		icon.SetPosition(q.x+questOffsetX, q.y+questOffsetY+iconOffsetY)
-
+		icon.SetPosition(x+questOffsetX, y+questOffsetY+iconOffsetY)
 		wg.AddWidget(icon)
-		wg.AddWidget(socket)
-		wg.AddWidget(button)
 	}
+
+	for i := 0; i < questsInAct; i++ {
+		currentQuest := i
+		buttons[i].OnActivated(func() {
+			var err error
+			for j := 0; j < questsInAct; j++ {
+				err = sockets[j].SetCurrentFrame(0)
+				if err != nil {
+					s.Error(err.Error())
+				}
+			}
+			if act-1 == s.selectedTab {
+				err = sockets[currentQuest].SetCurrentFrame(1)
+				if err != nil {
+					s.Error(err.Error())
+				}
+			}
+			s.onQuestClicked(currentQuest + 1)
+		})
+	}
+
+	for _, s := range sockets {
+		wg.AddWidget(s)
+	}
+
+	for _, b := range buttons {
+		wg.AddWidget(b)
+	}
+
 	wg.SetVisible(false)
 
 	return wg
 }
 
-func (s *QuestLog) makeQuestCallback(n int) func() {
-	return func() {
-		s.onQuestClicked(n + 1)
+func (s *QuestLog) makeQuestIconForAct(act, n int) (*d2ui.Sprite, error) {
+	icon, err := s.uiManager.NewSprite(s.questIconsTable(act, n), d2resource.PaletteSky)
+	if err != nil {
+		s.Error(err.Error())
+	}
+
+	switch s.questStatus[s.cordsToQuestID(act, n)] {
+	case d2enum.QuestStatusCompleted:
+		err = icon.SetCurrentFrame(completedFrame)
+	case d2enum.QuestStatusCompleting:
+		// that's not complet now
+		err = icon.SetCurrentFrame(0)
+		if err != nil {
+			s.Error(err.Error())
+		}
+
+		icon.PlayForward()
+		icon.SetPlayLoop(false)
+		err = icon.SetCurrentFrame(completedFrame)
+		s.questStatus[s.cordsToQuestID(act, n)] = d2enum.QuestStatusCompleted
```

Comment:
Has anybody some idea, how to do that?
I suppose, that for now animation starts before we open quest log, and we don't see it.